napari/layers/utils/layer_utils.py

```diff
@@ -684,58 +684,201 @@ def features_to_pandas_dataframe(features: Any) -> pd.DataFrame:
+        DataFrame index is used.
     """
-    if properties is not None or property_choices is not None:
-        features = _features_from_properties(
-            properties=properties,
-            property_choices=property_choices,
-            num_data=num_data,
+
+    def __init__(
+        self,
+        values: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
+        *,
+        num_data: Optional[int] = None,
+    ):
+        self._values = _validate_features(values, num_data=num_data)
+        self._defaults = self._make_defaults()
+
+    @property
+    def values(self) -> pd.DataFrame:
+        """The feature values table."""
+        return self._values
+
+    def set_values(self, values, *, num_data=None):
+        """Sets the feature values table."""
+        self._values = _validate_features(values, num_data=num_data)
+        self._defaults = self._make_defaults()
+
+    def _make_defaults(self) -> pd.DataFrame:
+        """Makes the default values table from the feature values."""
+        return pd.DataFrame(
+            {
+                name: _get_default_column(column)
+                for name, column in self._values.items()
+            },
+            index=range(1),
+            copy=True,
         )
-    else:
-        features = _validate_features(features, num_data=num_data)
-    defaults = pd.DataFrame(
-        {
-            name: _get_default_column(column)
-            for name, column in features.items()
-        },
-        index=range(1),
-        copy=True,
-    )
-    return features, defaults
+
+    @property
+    def defaults(self) -> pd.DataFrame:
+        """The default values one-row table."""
+        return self._defaults
+
+    def properties(self) -> Dict[str, np.ndarray]:
+        """Converts this to a deprecated properties dictionary.
+
+        This will reference the features data when possible, but in general the
+        returned dictionary mary contain copies of those data.
+
+        Returns
+        -------
+        Dict[str, np.ndarray]
+            The properties dictionary equivalent to the given features.
+        """
+        return _features_to_properties(self._values)
+
+    def choices(self) -> Dict[str, np.ndarray]:
+        """Converts this to a deprecated property choices dictionary.
+
+        Only categorical features will have corresponding entries in the dictionary.
+
+        Returns
+        -------
+        Dict[str, np.ndarray]
+            The property choices dictionary equivalent to this.
+        """
+        return {
+            name: series.dtype.categories.to_numpy()
+            for name, series in self._values.items()
+            if isinstance(series.dtype, pd.CategoricalDtype)
+        }
+
+    def currents(self) -> Dict[str, np.ndarray]:
+        """Converts the defaults table to a deprecated current properties dictionary."""
+        return _features_to_properties(self._defaults)
```

Comment:
```suggestion
        return _features_to_properties(self.defaults)
```

Reply:
I don't feel strongly, but in general I prefer using the attribute directly, rather than going through the property. Any objections if I keep this?

Reply:
No objections. Most of my comments are also questions, You wrote the code so I trust you.